[string] String(decoding:as:) fast path for withContiguousStorageIfAvailable

milseman opened the pull request:
Switch String(decoding:as) and other entry points to call
withContiguousStorageIfAvailable rather than use _HasContiguousBytes.

<rdar://problem/59148099>, SR-12125

rdar://problem/59148099

milseman commented:
@swift-ci please benchmark

Catfish-Man reviewed (approved):
Wonderful!

milseman committed:
[gardening] Delete Trailing Whitespace

milseman committed:
[string] _HasContiguousBytes -> withContiguousStorageIfAvailable

Switch String(decoding:as) and other entry points to call
withContiguousStorageIfAvailable rather than use _HasContiguousBytes.

milseman commented:
@swift-ci please benchmark

weissi reviewed (approved):
Awesome, thank you! @milseman shouldn't we add a test thought that validates we're doing the right thing?

milseman commented:
@weissi 

> shouldn't we add a test thought that validates we're doing the right thing?

There shouldn't be observable behavior differences from this change, just performance. But I do agree we should add a benchmark. I'll work on that.

weissi commented:
@milseman what about a LIT test like the ones I wrote when I last made sure more things are hitting the fast paths? https://github.com/apple/swift/pull/21743/files#diff-ba4d75f225736a97fc5a116d6ecf0821R1

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please benchmark

milseman commented:
> what about a LIT test like the ones I wrote when I last made sure more things are hitting the fast paths?

Doh, I think you're right. The benchmarks I added are probably still useful (especially the non-contiguous one and the contrast between the two), but a check on the SIL is nicer to enforce this.

milseman commented:
That's a little all over the place.

@swift-ci please benchmark

milseman commented:
Ugh, so with this change, release-no-assertions pass the test, but release-with-assertions have different IRGen behavior. Calls are now denoted as `tail` calls. Most vexingly, 

```
define swiftcc { i64, %swift.bridge* } @"$s22utf8_decoding_fastpath16decodeUMBPAsUTF8ySSSrys5UInt8VGF"(i64 %0, i64 %1) #0 {
  %3 = tail call swiftcc { i64, %swift.bridge* } @"$s22utf8_decoding_fastpath15decodeUBPAsUTF8ySSSRys5UInt8VGF"(i64 %0, i64 %1) #0
  ret { i64, %swift.bridge* } %3
}
```

Meaning that in assertion builds, the UMBP version just calls the UMB version (which is actually fine and a good idea, it's probably that there's an assertion preventing some other optimization from messing this up).

milseman commented:
... and code size regressions for the benchmark file is not great, since that will be conflated with the new benchmarks I'm adding. I'm going to spin those off because I think I have a change that will also improve code size.

atrick commented:
@milseman you could try converting those tests to .sil to see if that helps. I'm not sure why they were .ll tests to begin with.

I personally don't think the benchmark code size numbers are meaningful. They might correlate with real code size or might just be random. SCK code size is relevant. I think "swift-ci compiler performance" or something like that gets you aggregate code size for the SCK. We don't have any way to get a per project breakdown without running it locally. This takes time, which is why I haven't checked in anything that significantly perturbs codegen in the past year!

weissi commented:
@milseman awesome. Definitely agree that the benchmarks are super useful but SIL checks are just more stable. Regarding the tail calls to other functions: I've added `-swiftmergefunc-threshold=0` to get around that in my tests, then it doesn't merge similar functions anymore.

milseman commented:
@atrick It's more like I saw that both the fast-path and the path that makes an Array out of the code units were always-inline, and Array initializer loops is probably also always inline. I haven't checked the disassembly, but this would be tragic if so.

atrick commented:
@milseman I think it's a good idea to hand-outline the slow paths.

If you see problems where always-inline functions are inlined into slow paths, I think that should be fixed by adding an `@inline(_optimize)` attribute, which would not inline into slow paths, but would only override size-based heuristics, then migrating most of the stdlib functions to that annotation instead. That requires first adding parser support, then when everyone has the new compiler starting to use the attribute.

milseman commented:
@swift-ci please benchmark

milseman committed:
[string] Outline cold path from initializer

Outline the cold, non-contiguous UTF-8 path from String(decoding:as:),
saving ~40 bytes (33%) of code size (x86_64 and arm64) from every call
site where the contiguity check cannot be constant folded away.

milseman commented:
@swift-ci please benchmark

milseman commented:
For those following along, migrating these test cases to SIL is pending @atrick's SIL optimizer changes. The optimizer in unable to constant-fold the metatype comparison, and thus the entirety of the code is present in SIL form (LLVM later optimizes this). It is important to fold that comparison at the SIL level to accurately represent the actual program being compiled. Since 99+% of the time these generics are known at compilation time, failing to eliminate the branch to the slow path inside this always-inline function grossly distorts any size-based heuristics.

milseman commented:
Update: UnsafeRawBufferPointer (and mutable, and slices) cannot implement withContiguousStorageIfAvailable as that would create a typed pointer from untyped memory.

The longer-term solution to avoid having two checks is to add a withContiguousRawStorageIfAvailable that runs a closure over UnsafeRawBufferPointer. It can have a default implementation that calls withContiguousStorageIfAvailable and constructs the raw pointer from that (so existing wCSIA clients get this new path for free). Then, we can run only over the new API.

For now, I'll have to keep the `_HasContiguousBytes` call and just add a `wCSIA` fast path. Luckily with @atrick's SIL peephole optimizations, that can fold away at the SIL level. `_HasContiguousBytes` is a known closed set of conformances so that can (in theory) fold away. The only downside would be a generic function that calls `String(decoding:as:)` would have code size for both paths, but luckily that isn't too large, and is made up for by me outlining the slow Array-creating path.

karwa commented:
Yes, that’s what I implemented in https://github.com/apple/swift/pull/22028

Unfortunately, basically the entire internal (ABI-exposed) String API works in terms of typed pointers. So supporting URBP would require duplicating all of those entry points.

milseman commented:
@karwa Doh, I forgot about that. I like most of the changes in that PR, but we'll want to do a little extra to preserve the ABI. For this PR I'll try to add the `wCSIA` fast-path.

> Unfortunately, basically the entire internal (ABI-exposed) String API works in terms of typed pointers. So supporting URBP would require duplicating all of those entry points.

We can change that by just having them remove the typed-ness and call raw internal functions. It is always type safe to go from typed to untyped, it's the reverse that's tricky.

> Even though _HasContiguousBytes vends a raw pointer (so URBP can conform), the String initialiser makes a typed pointer using assumingMemoryBound, which might be an invalid binding. So the old code technically has UB, if I understand memory binding correctly (which I might not).

This would expose the potential for UB if there was an existing type binding for that memory, that is if there were any typed pointers that weren't over `UInt8`. IIUC (please correct me @atrick), additionally those types pointers would have to be non-trivial in this specific case, because the code we're running is self contained (no aliasing potential).

So in this case, since the set of `_HasContiguousBytes` is closed, the one UB scenario would be: if a caller of `String(decoding:as:)` were to pass in a URBP that held non-trivially-bound memory. So we should probably fix this by changing our ABI entry points to be over raw memory. We should add a public `withRawContiguousStorageIfAvailable` (as you did privately in your PR) going through SE. We'll want to do this in an ABI preserving fashion.

Let me get this performance optimization in now, and we can try to get the right ABI in place.

atrick commented:
TLDR: @milseman's plan is good

Ultimately, I think it would be better if the character decoding utilities did not require typed pointers. Ideally, there would be a different abstraction over a raw buffer and an encoding type.

However, that is not necessary in order to avoid undefined behavior. Internally, the String implementation can expose a typed pointer for the duration of a closure. As long as we know that arbitrary code does not access the same memory using a pointer with a different type, then it's safe. The String implementation can guarantee this because it does not execute any arbitrary user code during decoding. To make this 100% safe, we eventually need some compiler support (`URBP.withMemoryRebound`), but for now we can use `assumingMemoryBound(to:)` as a placeholder. It will be safe within reason here for various reasons. Eventually, I'll vet those 
uses of `assumingMemoryBound(to:)` and convert them into a `URBP.withMemoryRebound`.

Sorry, I don't have time to show example code, but hopefully you get the idea. The important thing is that we have multiple viable approaches to the String implementation. The real question is how to allow 3rd party Collections to participate in fast character decoding, and other fast paths the require contiguous storage.

There has never been any question in my mind that we need something like `withContiguousRawStorageIfAvailabe` (or `withBytesIfAvailable`) as a lowest common denominator to handle any byte-buffer-like collection. **Note: Any upcoming `ByteBuffer` APIs will need this.** If we had to choose between the raw form and the existing wCSIA, we would need to choose the raw form. We can always create a typed view over raw bytes, we just can't use the strictly typed UnsafePointer/UBP to do that. On the other hand any collection that already implements wCSIA will automatically provide the raw form via a default implementation.

milseman committed:
[string] Restore _HasContiguousBytes for untyped storage

UnsafeRawBufferPointer cannot implement
withContiguousStorageIfAvailable because doing so would potentially
create a typed pointer from untyped data.

milseman commented:
@swift-ci please benchmark

milseman commented:
Interesting that restoring the `_HasContiguousBytes` check dropped the Data benchmark gains. 

@Catfish-Man, it seems highly likely that the check isn't getting eliminated, even at the LLVM IR level. That might be an interesting thing to explore for some easy Data->String wins.

milseman commented:
The style of the benchmarks make them frustratingly hard to investigate to crawl through the disassembly, but I found it does not remove the check:

```
_$s14DataBenchmarks6string_4fromySi_10Foundation0A0VtF
...
0000000000008f75         call       _$s10Foundation4DataV15_RepresentationOWOy  ; _$s10Foundation4DataV15_RepresentationOWOy
0000000000008f7a         lea        rdi, qword [_$ss19_HasContiguousBytes_pMD]  ; argument #1 for method ___swift_instantiateConcreteTypeFromMangledName, _$ss19_HasContiguousBytes_pMD
0000000000008f81         call       ___swift_instantiateConcreteTypeFromMangledName ; ___swift_instantiateConcreteTypeFromMangledName
0000000000008f86         mov        r8d, 0x6
```

milseman commented:
Ugh, so now that I've restored the `_HasContiguousBytes` check (and moved it below the wCSIA locally so that can get constant folded), I'm back to the case where `decodeUBPAsUTF8` in the SIL representation's wCSIA closure won't get inlined despite everything being always_inline and frankly trivial.

I sprinkled the magic `Builtin.onFastPath()` into the closure call to hint to the optimizer that it should really inline this fairly trivial closure. This used to work, but with the restored `_HasContiguousBytes` that is no longer happening, despite that check being optimized away.

@atrick do you know if there is a pass ordering problem I'm hitting?

atrick commented:
@milseman I'm not sure what `decodeUBPAsUTF8` is. I don't think any `@semantic` calls are involved, so I can't think of a known pass pipeline problem. If a function marked `@inline(always)` is not being inlined, it sounds like the optimizer should be fixed ASAP because that's your last resort. You should file a bug, but if you need it to be fixed before merging this PR we should find someone who has time to look at the `-Xllvm -debug-only=sil-inliner` output.

milseman commented:
Currently working around the issue that Substring's wCSIA closure is not being inlined, despite the `Builtin.onFastPath` call.

milseman committed:
[string] Move wCSIA check higher than _HasContiguousBytes

milseman committed:
[string] Convert IR tests to SIL tests

milseman commented:
@swift-ci please test

milseman commented on test/SILOptimizer/utf8_decoding_fastpath.swift:
```diff
@@ -12,101 +12,199 @@
+//
+// NOTE: The SIL optimizer cannot currently fold away a (UTF16.self ==
+// UTF8.self) metatype comparison, so we have to disabel the check-not for UTF-8
+// construction :-(
```
@atrick should metatype inequality be foldable as well?

atrick replied:
It should be easy, just didn't fall out naturally from handling equality. You have to also check that there's no potential relationship between the type values. If you file a bug I or someone else can circle around to it.
